### PR TITLE
New resource: aws_lambda_invocation

### LINKED ISSUE #4746: execute aws_lambda_invocation only once
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

After creating an aws_lambda_function I want directly to trigger it for a first initialization. Currently I'm doing this using a provisioner and and some shell commands, which are invoking the the lambda with the AWS CLI.

I've seen there is now an aws_lambda_invocation data source, which does exactly the same like my provisioner. In my case, the disadvantage of a data source is, that the lambda is executed every time I do 
a "terraform plan" or "terraform apply", which is not needed and increases the execution time (and AWS costs).

It would be nice to be able to trigger a lambda function only once

### New or Affected Resource(s)
Option 1
Create:
```hcl
resource "aws_lambda_invocation" "invoke"
```
Option 2
Update:
```hcl
data "aws_lambda_invocation" "invoke"
```
### Potential Terraform Configuration

I would prefer the first option, to create a new ressource type. It would look like the data source:
```hcl
resource "aws_lambda_invocation" "example" {
  function_name = "${aws_lambda_function.lambda_function_test.function_name}"

  input = <<JSON
  {
    "key1": "value1",
    "key2": "value2"
  }
  JSON
}
```

As an alternative an optional attribute in the data source can manage the behavior:
```hcl
data "aws_lambda_invocation" "example" {
  function_name = "${aws_lambda_function.lambda_function_test.function_name}"
  trigger_once  = true

  input = <<JSON
  {
    "key1": "value1",
    "key2": "value2"
  }
  JSON
}
```

### References

No references

### PR DESCRIPTION
**Pull request by pascal236**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #4746

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccResourceAwsLambdaInvocation_basic'    
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccResourceAwsLambdaInvocation_basic -timeout 180m
=== RUN   TestAccResourceAwsLambdaInvocation_basic
=== PAUSE TestAccResourceAwsLambdaInvocation_basic
=== CONT  TestAccResourceAwsLambdaInvocation_basic
--- PASS: TestAccResourceAwsLambdaInvocation_basic (67.40s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       69.134s
```

```
$ make testacc TESTARGS='-run=TestAccResourceAwsLambdaInvocation_qualifier'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccResourceAwsLambdaInvocation_qualifier -timeout 180m
=== RUN   TestAccResourceAwsLambdaInvocation_qualifier
=== PAUSE TestAccResourceAwsLambdaInvocation_qualifier
=== CONT  TestAccResourceAwsLambdaInvocation_qualifier
--- PASS: TestAccResourceAwsLambdaInvocation_qualifier (70.58s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       72.298s
```

```
$ make testacc TESTARGS='-run=TestAccResourceAwsLambdaInvocation_complex'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccResourceAwsLambdaInvocation_complex -timeout 180m
=== RUN   TestAccResourceAwsLambdaInvocation_complex
=== PAUSE TestAccResourceAwsLambdaInvocation_complex
=== CONT  TestAccResourceAwsLambdaInvocation_complex
--- PASS: TestAccResourceAwsLambdaInvocation_complex (67.01s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       68.750s
```

```
$ make testacc TESTARGS='-run=TestAccResourceAwsLambdaInvocation_Triggers'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccResourceAwsLambdaInvocation_Triggers -timeout 180m
=== RUN   TestAccResourceAwsLambdaInvocation_Triggers
=== PAUSE TestAccResourceAwsLambdaInvocation_Triggers
=== CONT  TestAccResourceAwsLambdaInvocation_Triggers
--- PASS: TestAccResourceAwsLambdaInvocation_Triggers (158.56s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       160.289s
```

All credits to dedunumax #14388 (I am just fixing a stale PR)

### CODE REVIEW

File: aws/resource_aws_lambda_invocation.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validation.StringIsJSON,
+			},
+
+			"invoke_on_update": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  true,
+			},
```
Comment (DJAlPee) [inline aws/resource_aws_lambda_invocation.go:42]: Using `triggers` with a map might be more consistent with other terraform (AWS) resources (and more powerful):
https://github.com/hashicorp/terraform-provider-aws/blob/5283e38ff9fa019c0d8aa3059a41a03cf431a2c2/aws/resource_aws_api_gateway_deployment.go#L45-L50

If I got it right, there would be no further checking (like in #L88 below) necessary anymore.

Comment (pascal236) [reply aws/resource_aws_lambda_invocation.go]: Agree, triggers seems to be standard and also nicely integrated into the provider itself. I amended the PR and it is now using triggers. Also included a test.

Thanks for the review.

Comment (DJAlPee) [reply aws/resource_aws_lambda_invocation.go]: Perfect. I also like the updated example in the docs.

Comment (pascal236) [reply aws/resource_aws_lambda_invocation.go]: As everything seems fine, I will now resolve this conversation.

@DJAlPee Thanks for the review

**Comment by agseijas** [comment]
Dunno if it can be done without changes to this PR, what about changing the input on tf destroy phase?

**Comment by dedunumax** [comment]
@DJAlPee Thanks a lot for your swift review. Can we merge this before it gets stale again?

**Comment by dedunumax** [comment]
@DJAlPee Thanks a lot for your swift review. Can we merge this before it gets stale again?

**Comment by DJAlPee** [comment]
Sorry, but I'm not a member of Hashicorp/Terraform team (yet) 😄.
No idea, how to ping a maintainer...

**Comment by pascal236** [comment]
If anyone knows how to push/notify this to one of the terraform-aws reviewer, please let me know :)

**Comment by avivek** [comment]
Is this on the radar??
It will be absolutely useful for our workflows.

**Review by YakDriver — APPROVED** [review]
Looks great! 🎉 

Output of acceptance tests (`us-west-2`):
```console
% make testacc TESTS=TestAccLambdaInvocation_ PKG=lambda                     
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/lambda/... -v -count 1 -parallel 20 -run='TestAccLambdaInvocation_'  -timeout 180m
--- PASS: TestAccLambdaInvocation_basic (31.34s)
--- PASS: TestAccLambdaInvocation_qualifier (45.57s)
--- PASS: TestAccLambdaInvocation_complex (53.02s)
--- PASS: TestAccLambdaInvocation_triggers (68.51s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/lambda	69.753s
```

Output of acceptance tests (GovCloud):
```console
% make testacc TESTS=TestAccLambdaInvocation_ PKG=lambda
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/lambda/... -v -count 1 -parallel 20 -run='TestAccLambdaInvocation_'  -timeout 180m
--- PASS: TestAccLambdaInvocation_basic (33.75s)
--- PASS: TestAccLambdaInvocation_qualifier (40.24s)
--- PASS: TestAccLambdaInvocation_complex (46.88s)
--- PASS: TestAccLambdaInvocation_triggers (91.14s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/lambda	92.475s
```

**Pull request merged**